content/docs/projects/tdd.md

```diff
@@ -0,0 +1,602 @@
+
+The `tearDown` function is used to perform any necessary clean up tasks after unit tests within a `TestCase` class are run. This can include things like closing any open database connections, cleaning up temporary files, or restoring any modified global state.
+
+The `setUp` and `tearDown` functions are both optional, and you only need to use them if your tests require some kind of setup or cleanup. In most cases, it is a good idea to use these functions to make your tests more modular and easier to maintain.
+
+5. Once you have written your tests, you can run them using the `bench` command-line tool from within the app's directory. This will execute all of the test cases and display the results.
+
+```bash
+bench run-tests
+```
+
+> Note: Testing is disabled by default. You can enable tests by entering following command:
+
+```bash
+bench --site site_name set-config allow_tests true
+```
+
+That's it! These are the basic steps for writing and running unit tests in the Frappe framework. Keep in mind that there are many other features and options available for unit testing in Frappe, so you may want to consult the [official Frappe documentation](https://frappeframework.com/docs/v13/user/en/guides/automated-testing/unit-testing) for more information.
+
+### 5.2.4 Additional Reminders and Tips
+
+A few things to remember when writing unit tests:
+
+1. As much as possible, avoid using default document names (e.g., `_Test Item`). Create documents with names specific to the unit test to ensure that the attributes of the item are those known by the unit test. One way to do this is to use the test name when creating the document (e.g., `_Test Item test_deduct_item_discounts_from_item_amounts`).
+2. Avoid deleting documents and transactions to avoid errors when deleting documents linked to other documents.
+3. Start with a small, well-defined task. This will help you focus on the task at hand and avoid getting overwhelmed.
+4. Write a test for the task before you write the code. This will help you understand the requirements and constraints of the task, and it will give you a clear goal to work towards.
+5. Write the minimum amount of code needed to make the test pass. This will help you avoid over-engineering and keep your code simple and focused.
+6. Refactor your code to improve its design and readability. This will help you make your code more maintainable and easier to understand.
+7. Repeat the process for each new feature or task you want to implement. This will help you build a suite of tests that will give you confidence in the correctness and completeness of your code.
+8. Be persistent and keep practicing. Test-driven development can be challenging at first, but it becomes easier and more natural with time and practice.
```

Comment:
Might also be good to add the tip to create a separate site for running the tests and sharing the instructions on how to do that.

Reply:
856cf12959eea44b2307f78bbf6ce9b83a24a51b